packages/@aws-cdk/aws-cognito/lib/user-pool-domain.ts

```diff
@@ -0,0 +1,156 @@
+import { ICertificate } from '@aws-cdk/aws-certificatemanager';
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { AwsCustomResource, AwsCustomResourcePolicy, AwsSdkCall, PhysicalResourceId } from '@aws-cdk/custom-resources';
+import { CfnUserPoolDomain } from './cognito.generated';
+import { IUserPool } from './user-pool';
+
+/**
+ * Represents a user pool domain.
+ */
+export interface IUserPoolDomain extends IResource {
+  /**
+   * The domain that was specified to be created.
+   * If `customDomain` was selected, this holds the full domain name that was specified.
+   * If the `cognitoDomain` was used, it contains the prefix to the Cognito hosted domain.
+   * @attribute
+   */
+  readonly domainName: string;
+}
+
+/**
+ * Options while specifying custom domain
+ * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-add-custom-domain.html
+ */
+export interface CustomDomainOptions {
+  /**
+   * The custom domain name that you would like to associate with this User Pool.
+   */
+  readonly domainName: string;
+
+  /**
+   * The certificate to associate with this domain.
+   */
+  readonly certificate: ICertificate;
+}
+
+/**
+ * Options while specifying a cognito prefix domain.
+ * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-assign-domain-prefix.html
+ */
+export interface CognitoDomainOptions {
+  /**
+   * The prefix to the Cognito hosted domain name that will be associated with the user pool.
+   */
+  readonly domainPrefix: string;
+}
+
+interface UserPoolDomainTypeConfig {
+  readonly domain: string;
+
+  readonly certificate?: ICertificate;
+}
+
+/**
+ * The type of user pool domain. Two types are currently supported - custom domain and cognito domain.
+ */
+export class UserPoolDomainType {
```

Comment:
Name is a little long for my tastes.

I don't know man. Maybe it was fine the first time, with 3 optional fields and some runtime validation. It feels like a pretty complicated dance right now very very little benefit right now.

Do with that as you will, I don't feel super strongly.